File: go-version/leetcode/LC1863/subset_xor_sum.go
package LC1863

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/4/5 下午3:36
 * @FilePath: subset_xor_sum
 * @Description: 通过深度优先搜索遍历全部子集
 */

func subsetXORSum(nums []int) int {
	ans := 0
	n := len(nums)
	var dfs func(int, int, []int)

	dfs = func(val int, idx int, arr []int) {
		if idx == n { //递归终止条件
			ans += val
			return
		}
		// 选择当前数字
		dfs(val^arr[idx], idx+1, arr)
		// 跳过当前数字
		dfs(val, idx+1, arr)
	}

	dfs(0, 0, nums)
	return ans
}

File: python-version/LC1863/subset_xor_sum.py
class Solution:
    def __init__(self):
        self.__n = 0
        self.__ans = 0

    def __dfs(self, val: int, idx: int, nums: list[int]):
        if idx == self.__n:  # 递归终止条件
            self.__ans += val
            return
        # 选择当前数字
        self.__dfs(val ^ nums[idx], idx + 1, nums)
        # 跳过当前数字
        self.__dfs(val, idx + 1, nums)

    def subsetXORSum(self, nums: list[int]) -> int:
        self.__n = len(nums)
        self.__ans = 0
        self.__dfs(0, 0, nums)
        return self.__ans
